[ML] Ensure progress is complete by the time DFA job stops

dimitris-athanasiou opened the pull request:
Test failures in #74342 have shown that it is possible for a
data frame analytics job to complete and yet for its progress
to appear incomplete.

The reason this may happen is that we persist progress asynchronously
except for when we're marking the task complete. When we mark the task
complete, we make another try to persist complete progress. However,
we skip persistence if the existing progress document is the same.
Those previous fire-and-forget index requests may end up processed
in a different order resulting into overwriting a latter progress
document.

This commit fixes this by making the progress persist calls at
the end of each step wait for the index request to complete. 

Closes #74342

dimitris-athanasiou committed:
[ML] Ensure progress is complete by the time DFA job stops

Test failures in #74342 have shown that it is possible for a
data frame analytics job to complete and yet for its progress
to appear incomplete.

The reason this may happen is that we persist progress asynchronously
except for when we're marking the task complete. When we mark the task
complete, we make another try to persist complete progress. However,
we skip persistence if the existing progress document is the same.
Those previous fire-and-forget index requests may end up processed
in a different order resulting into overwriting a latter progress
document.

This commit fixes this by making the progress persist calls at
the end of each step wait for the index request to complete. In
addition, we abort persisting state if the task has been cancelled.
This is to avoid a progress document left behind if the job gets
force-closed and deleted in quick succession (see #61913).

Closes #74342

elasticmachine commented:
Pinging @elastic/ml-core (Team:ML)

benwtrent commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsTask.java:
```diff
@@ -187,8 +187,8 @@ public void setFailed(Exception error) {
+        persistProgress(client, taskParams.getId(), runnable);
     }
 
     // Visible for testing
```
I think it would be beneficial to log the "persisted progress" object in the `debug` logger. This would probably require a `setonce` or something to grab the `progressToStore` object.

dimitris-athanasiou replied:
We already log progress if we go on to persist it. I added the progress in the debug logging when we skip storage because it's the same as the previously stored progress.

benwtrent replied:
@dimitris-athanasiou I was actually thinking of logging on line 203: `LOGGER.debug("[{}] Successfully indexed progress document", jobId);`

Which progress document? There is no way to tie in the loglines the progress starting to be indexed and then the same one completing.

dimitris-athanasiou replied:
Ah, yes. Good point. Will add that too.

benwtrent reviewed (approved):
I think this change is good. I added a comment about extending logging a bit to make it easier to troubleshoot issues in the future.

dimitris-athanasiou committed:
Remove abort on cancelled plus log progress when same as previous

dimitris-athanasiou commented:
I have removed the abortion of storing progress when the task is cancelled as that is causing other test failures.

With regard to the issue of ending up with a progress document if the job gets force-stopped and deleted in quick succession I think the solution is to do what we already do for anomaly detection jobs. When a job is created, we check if there are left over documents matching that job and fail job creation. I think this is an acceptable and consistent way to handle this scenario. I will follow up in a different PR.

dimitris-athanasiou commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ensure-progress-is-complete-by-the-time-dfa-stops

dimitris-athanasiou commented:
@benwtrent Could you take another look please?

dimitris-athanasiou committed:
Log which progress was indexed

dimitris-athanasiou commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ensure-progress-is-complete-by-the-time-dfa-stops

dimitris-athanasiou commented:
run elasticsearch-ci/part-2

benwtrent reviewed (approved):
:shipit:
